### PR TITLE
#URL macro

### PR DESCRIPTION
**Pull request by mallexxx**
Task URL: https://app.asana.com/0/42792087274227/1206542455948401/f
BSK PR: https://github.com/duckduckgo/BrowserServicesKit/pull/657
macOS PR: https://github.com/duckduckgo/macos-browser/pull/2179
iOS PR: https://github.com/duckduckgo/iOS/pull/2540

### CODE REVIEW

File: Sources/Macros/MacroDefinitions.swift [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+/// Compile-time validated URL instantiation
```
Comment (ayoy) [inline Sources/Macros/MacroDefinitions.swift]: @mallexxx I think we should work on the documentation of #URL macro a little bit before it gets released. From the user perspective I've found a few things that weren't obvious:
1. Classic `http://💩.la` is not valid
2. File URLs seem to be out of scope of this macro (but a URL with `file://` scheme works)

Could you please update the docs to provide more information about what to expect from the macro, when to not use it, and any further information about it that you consider important?

Should we consider adding a swiftlint rule to warn about `URL.init(string:)` initializer use?

Comment (mallexxx) [reply Sources/Macros/MacroDefinitions.swift]: the idea is to disable any applied URL conversions like punycode or no-scheme urls, only 1-1 mapped valid URLs, I‘ll add some docs

Of course I considered adding some swiftlint rule but it seems more of a task for a static analyzer as trying to apply a regular regex rule hogs the build time too much

Comment (ayoy) [reply Sources/Macros/MacroDefinitions.swift]: > the idea is to disable any applied URL conversions like punycode or no-scheme urls, only 1-1 mapped valid URLs

This already sounds great (as part of the documentation) - makes it clear when the macro can be used. So the bottom line is, we still consider `URL.init(string:)` to still have valid use cases, such as when translating user input to URL, etc. In this case I don't think we even need to think about the static analysis. Thanks for clarifying it!

**Comment by mallexxx** [comment]
Updated the comment and opened iOS PR

File: Sources/Macros/MacroDefinitions.swift [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+/// Usage: `let url = #URL("https://duckduckgo.com")`
+///
+/// - Note: Strings like "http://💩.la" or "1" are not valid #URL parameter values.
+/// To instantiate a parametrized URL use `URL.appendingPathComponent(_:)` or `URL.appendingParameters(_:allowedReservedCharacters:)`
+/// To instantiate a URL from a String format, use `URL(string:)`
+///
+/// - Parameter string: valid URL String Literal with URL scheme and
+/// - Returns: URL instance if provided string argument is a valid URL
+/// - Throws: Compile-time error if provided string argument is not a valid URL
+///
```
Comment (ayoy) [inline Sources/Macros/MacroDefinitions.swift]: ❤️

**Review by ayoy — APPROVED** [review]
Great job @mallexxx ! 💪

**Pull request merged**